src/events/tls.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for related applications on the `certificates` relation interface."""
+import base64
+import logging
+import re
+from typing import TYPE_CHECKING
+
+from charms.tls_certificates_interface.v3.tls_certificates import (
+    CertificateAvailableEvent,
+    TLSCertificatesRequiresV3,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import ActionEvent, RelationCreatedEvent, RelationJoinedEvent
+from ops.framework import EventBase, Object
+
+if TYPE_CHECKING:
+    from charm import OpensearchDasboardsCharm
+
+logger = logging.getLogger(__name__)
+
+
+class TLSEvents(Object):
+    """Event handlers for related applications on the `certificates` relation interface."""
+
+    def __init__(self, charm):
+        super().__init__(charm, "tls")
+        self.charm: "OpensearchDasboardsCharm" = charm
+        self.certificates = TLSCertificatesRequiresV3(self.charm, "certificates")
+
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_created"),
+            self._on_certs_relation_created,
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_joined"), self._on_certs_relation_joined
+        )
+        self.framework.observe(
+            getattr(self.certificates.on, "certificate_available"), self._on_certificate_available
+        )
+        self.framework.observe(
+            getattr(self.certificates.on, "certificate_expiring"), self._on_certificate_expiring
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_broken"), self._on_certs_relation_broken
+        )
+
+        self.framework.observe(
+            getattr(self.charm.on, "set_tls_private_key_action"), self._set_tls_private_key
+        )
+
+    def _on_certs_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for `certificates_relation_created` event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.charm.state.cluster.update({"tls": "enabled", "switching-encryption": "started"})
+
+    def _on_certs_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handler for `certificates_relation_joined` event."""
+        if not self.charm.state.cluster.tls:
```

Comment:
What is the goal of informing the cluster you have started TLS or not? You can already discover that by checking the len(...`get_relation(TLS_RELATION)`) > 0 (i.e. you have to setup TLS.

That spares the charm from an unnecessary defer.

Reply:
I see your point.

So let me do this.

I open a separate ticket for TLS workflow reveiw, that's to be addressed before the [Final Cleanup](https://warthogs.atlassian.net/browse/DPE-3954), dedicated to the TLS workflow.

Most of these decision were literally copied from the ZK code, where additinal logic and measures have to be applied for quorum and stuff. We can likely get rid of much of that however first I want to understand why they are in place.
As --equally-- there may be issues encountered, leading to this code, that we may want to consider as well.
(Typically, as for the above... I could imagine issues around response on the existence of the relation against the existence of a flag in the databag...? I don't know, but I'd like to investigate separately.)

I created https://warthogs.atlassian.net/browse/DPE-4033 with all your points to be addressed there.
(Feel free to check if I may be missing anything.)